### ISSUE TITLE
Atomic operations and constraint checks

### ISSUE THREAD

**Issue by jere-software** [Ada-Rapporteur-Group/UCI-Prototype#114]
I had a question about the intent of System.Atomic_Operations.Integer_Arithmetic.  I started using them only to find something unexpected:  they don't actually do the range checks (in GNAT at least).  Here was my test program:

```Ada
pragma Ada_2022;

with Ada.Text_IO; use Ada.Text_IO;
with System.Atomic_Operations.Integer_Arithmetic;

procedure jdoodle is
    type A is new Integer with Atomic;
    package Ops is new System.Atomic_Operations.Integer_Arithmetic(A);
    
    v : aliased A := A'Last;
begin
    Ops.Atomic_Add(V,1);
    Put_Line(v'Image);
end jdoodle;
```

with Output:
```
-2147483648
```

My first thought was a compiler bug, but I did go check the RM page on this package and noticed that the RM specifies all the operations as Intrinsic (see my comments below to explain why this changes my perspective), so now I am not sure the compiler is in error.  So I wanted to ask:

Is it intended at the language level that atomic operations not check constraints when doing their operations?  

The RM does not seem to specify what the intent is on this and given that there are separate packages for modular vs integer (vs a singular generic for all discrete types with `"+"` and `"-"` defined), my intuition was that the integer ones would do range checks.  That said, when I researched atomic operations, I found that the implementations I found didn't do constraint checks either since they often directly map to an assembly instruction (hence why the intrinsic specification had me pause my train of thought).

I know it's fairly trivial to implement them myself using the Exchange package, but I wanted to see what the language defined intent was.  Also, if the intent is that the constraint checks are not to be performed, would it be worthwhile adding some clarification verbiage there in the RM so folks are not surprised.  In essence, right now, using Atomic_Add and similar is very close to doing an unchecked operation but a naive person (like myself) might not realize that based on the RM and package naming.

EDIT:  Side note, I couldn't figure out how to attach the "question" label to this, so my apologies.

**Comment by sttaft** [comment]
Range/Overflow checks are required to be performed unless they are suppressed.  This is true in general unless the RM says otherwise explicitly.

At one point, some versions of GNAT suppressed certain checks by default, so you should be sure that checks aren't being suppressed (you could explicitly Unsuppress them to be sure).

**Comment by jere-software** [comment]
I double checked with all checks unsuppressed (using GNAT 13.2 which is what I have available here), but the result is the same.  I agree with what you said here and it was my initial intuition:

> Range/Overflow checks are required to be performed unless they are suppressed. This is true in general unless the RM says otherwise explicitly.

But I wasn't sure because of the Intrinsic declaration (and the RM references mapping directly to machine instructions, which don't range check).  But you are saying it should be constraint checking, so that clears it up for me personally though I am not sure the RM feels as clear to me.

For reference, C.6.1 does say:

> The subprograms of this subsystem are Intrinsic subprograms (see [6.3.1](http://www.ada-auth.org/standards/22rm/html/RM-6-3-1.html)) in order to provide convenient access to machine operations that can provide these capabilities if they are available in the target environment.

**Comment by ArnaudCharlet** [comment]
> Range/Overflow checks are required to be performed unless they are suppressed. This is true in general unless the RM says otherwise explicitly.
> 
That makes sense at the RM level but not at the implementation level: what does it mean to perform an atomic addition with overflow check?
> At one point, some versions of GNAT suppressed certain checks by default, so you should be sure that checks aren't being suppressed (you could explicitly Unsuppress them to be sure).
> 

That’s not the issue here, gnat doesn’t know how to perform overflow checks on top of atomic intrinsics (which would no longer be intrinsic and would be way too inefficient to be useful since these intrinsics are typically used to implement efficient lock free algorithms)

Arno

**Comment by sttaft** [comment]
>> Range/Overflow checks are required to be performed unless they are suppressed. This is true in general unless the RM says otherwise explicitly.

> That makes sense at the RM level but not at the implementation level: what does it mean to perform an atomic addition with overflow check?

Good point.  We should recognize this issue in the RM somehow.  We could perhaps require that an exception be raised, but we couldn't require that the update not occur.

>> At one point, some versions of GNAT suppressed certain checks by default, so you should be sure that checks aren't being suppressed (you could explicitly Unsuppress them to be sure).

> That’s not the issue here, gnat doesn’t know how to perform overflow checks on top of atomic intrinsics (which would no longer be intrinsic and would be way too inefficient to be useful since these intrinsics are typically used to implement efficient lock free algorithms) Arno

Makes sense.  Probably should turn this query into an AI.

**Comment by jere-software** [comment]
Does this also mean (pending the results of the AI) that the atomic arithemetic operations be considered unchecked programming or is that too strong an assertion?  I ask because current possible implementations may not be able to do the constraint check and the value does get promoted incorrectly.  This also applies to the modular version if a constrained modular subtype is provided:

```Ada
type M is mod 2**32;
subtype S is M range 0 .. 2**16;  -- if this is supplied as the type to Moduluar_Arithmetic, the constraint check won't be performed either
```

**Comment by jere-software** [comment]
As an aside for anyone seeing this discussion later and looking for a workaround:

When I looked it up, the suggested work around to get range checking (for any language that has atomic ops) was to use a loop with compare and exchange.  Something similar to:

```Ada
    Item  : aliased Atomic_Type := Atomic_Type'Last;
    Value : Atomic_Type := 1;
    Temp  : aliased Atomic_Type := Item;
begin
    loop
        exit when Ops.Atomic_Compare_And_Exchange(Item, Temp, Temp + Value);
    end loop;
```
Which will catch the constraint error

I'm guessing this can't be intrinsic and also not every application can tolerate looping (even if the practical effect might be bounded and not infinite).  I'd guess in those scenarios a different tailored solution would have to be determined.

**Comment by jere-software** [comment]
> > That makes sense at the RM level but not at the implementation level: what does it mean to perform an atomic addition with overflow check?
> 
> Good point. We should recognize this issue in the RM somehow. We could perhaps require that an exception be raised, but we couldn't require that the update not occur.

Question on this:  Can this actually be done in this way?  If you do the constraint check after the atomic operation, how do you know that the value 0 was obtained via an addition that overflowed or was the addition ok and another task/thread changed the value to 0 through assignment "between" the atomic addition and the constraint check.

Task A:
```Ada
Item := 0;
```
Task B:
```Ada
Ops.Atomic_Add(Item,1);
-- what if Task A executes between the intrinsic call and the constraint check?
```

I'm not super experienced in atomic operations, so I apologize if this is a naive question.

**Comment by sttaft** [comment]
> I'm guessing this can't be intrinsic and also not every application can tolerate looping (even if the practical effect might be bounded and not infinite). I'd guess in those scenarios a different tailored solution would have to be determined.

In fact, I believe more machines have a compare-and-exchange instruction than have an atomic add, so many atomic add operations are implemented with a loop, just as you illustrated.  I think the challenge is that the "intrinsics" provided by the GCC backend try to abstract away this distinction, and those intrinsics don't perform overflow checking, and might guarantee wraparound.

I think the RM should clarify the situation.  Given the workaround to use compare-and-swap, it seems that guaranteeing wraparound might be the right choice.  As far as range constraints, there should probably still be a constraint check somewhere, even if it is after the atomic operation has occurred.  We don't want the value of a range-constrained object to be silently set to a value outside of its range.

This issue has been debated in the Rust community in the past: https://github.com/rust-lang/rust/issues/34618

**Comment by ARG-Editor** [comment]
Tuck replied to Arno:
 
>> That makes sense at the RM level but not at the implementation level:
what does it 
>>  mean to perform an atomic addition with overflow check?

> Good point. We should recognize this issue in the RM somehow. We could
perhaps
> require that an exception be raised, but we couldn't require that the
update not occur.

I think Tucker has this right, but let me clarify this for my own mind.

Ada has a rule that if an exception is raised, the target of an operation is
not modified. That rule doesn't make sense for an atomic operation, because
the whole point is to modify the object indivisibly. And machine
instructions are going to do that modification regardless of overflow.

OTOH, getting the wrong answer without any sort of notification is wrong
(assuming checks aren't suppressed). And there doesn't seem to be an
implementation problem checking overflow/range after the atomic operation
(the overflow bit will be set by the operation, so it can be tested, and the
range is an explicit check anyway).

We don't want to require the loop implementation (which could have the RM
correct effect), in part because it would add overhead exactly at a place
that we don't want it.

In practice, many of these super-efficient algorithms would be running with
checks suppressed anyway (hopefully they'd proved non-overflow in some other
way), so the main effect would be to protect the casual user from
themselves. Which is always the "Ada way".

It does sound like an AI is required, because requiring a loop
implementation would defeat part of the purpose of defining these
operations. But users should be able to detect the problem and take
corrective action when that is warranted, so simply having no checks seems
wrong.

               Randy.

**Comment by jere-software** [comment]
> And there doesn't seem to be an implementation problem checking overflow/range after the atomic operation (the overflow bit will be set by the operation, so it can be tested, and the range is an explicit check anyway).

What about subtypes with static predicates though?  The overflow bit is safe due to context switching but any value checking could be thrown off by being preempted by another task in between the atomic add machine instruction and the constraint check?  Or is there a way to guarantee that the OS doesn't preempt between the atomic_add and the static predicate check?

I know this is probably unlikely someone would use atomic_add with a type that has a static predicate more than just a range check, but it is still possible via the language, so maybe worth consideration potentially.

**Comment by ARG-Editor** [comment]
Jere asks: 

>I don't know if you saw my question above on this, but wanted to ask if
this
>is strictly true? Can the following ever occur:
>
>	1.	Task 1 issues the atomic add machine instruction and sets
the overflow bit 
>	2.	Task 2 runs and sets the value to 0, then later adds a value
to it 
>	3.	The OS gets back to task 1 and then checks the overflow bit
which
>           should be clear if the task 2 didn't overflow 
>
>	or is there a way to guarantee the OS cannot preempt between the
atomic add
> and the constraint check?

Preemption should be irrelevant; it has to preserve/restore the machine
state (the flags in particular). To see why, consider the normal case (no
atomics anywhere).

If we have a statement like:

    if A = 0 then

on an Intel machine, we might get code like:

    Mov EAX, [A]
    Cmp EAX, 0
    Jnz Else-Part

If preemptions did not preserve the machine flags, this common code would
not work if the task was preempted between the Cmp and the Jnz (the Jnz
depending upon the flags set by the Cmp). The preformance impact of having
to protect all comparsons from preemption would be prohibitive, it's way
easier to save the flags when preempting (all machines I know of have a way
to do this).

Indeed, the same sort of argument can be made for any standard part of the
machine state; the compiler shouldn't have to worry that some usual
operation might be preempted. So the scenario you outline would be wrong. If
there was an explicit Yield (or blocking), then all bets are off, but those
are under control of the compiler/user - one doesn't put those between
operations that care about the machine state.

                   Randy.

**Comment by jere-software** [comment]
> Preemption should be irrelevant; it has to preserve/restore the machine state (the flags in particular). To see why, consider the normal case (no atomics anywhere). If we have a statement like: if A = 0 then on an Intel machine, we might get code like: Mov EAX, [A] Cmp EAX, 0 Jnz Else-Part If preemptions did not preserve the machine flags, this common code would not work if the task was preempted between the Cmp and the Jnz (the Jnz depending upon the flags set by the Cmp). The preformance impact of having to protect all comparsons from preemption would be prohibitive, it's way easier to save the flags when preempting (all machines I know of have a way to do this). Indeed, the same sort of argument can be made for any standard part of the machine state; the compiler shouldn't have to worry that some usual operation might be preempted. So the scenario you outline would be wrong. If there was an explicit Yield (or blocking), then all bets are off, but those are under control of the compiler/user - one doesn't put those between operations that care about the machine state. Randy.

Yeah, I realized that before you replied, but my edit wasn't quick enough.  I did note though that maybe types with static predicates (unlikely but possible in the language) might still have trouble.  Context switches will save the overflow bits, but if the static predicate requires a value check, is it still a problem.  For referenence my edit was:

>What about subtypes with static predicates though? The overflow bit is safe due to context switching but any value checking could be thrown off by being preempted by another task in between the atomic add machine instruction and the constraint check? Or is there a way to guarantee that the OS doesn't preempt between the atomic_add and the static predicate check?

>I know this is probably unlikely someone would use atomic_add with a type that has a static predicate more than just a range check, but it is still possible via the language, so maybe worth consideration potentially.

**Comment by Jeff-Cousins** [comment]
The last time I looked at the source code generated by GNAT (a long time ago!), it didn't make use of the overflow flag, even though most common hardware provides one, but used some relatively long winded algorithm to test for overflow so as to be portable across all platforms.

Jeff

**Comment by ARG-Editor** [comment]
AI22-0125-1 has been constructed for this issue.

**Comment by ARG-Editor** [comment]
AI22-0125-1 was approved at ARG Meeting #63G, so this issue is completed.
